### DISCUSSION TITLE
extracting page values directly from paginated links

### DISCUSSION THREAD

**Discussion by benkoshy** [benkoshy/pagy#58] [Ideas]
The current mode when paginating from complex search results using POST request is by overriding `pagy_url_for` and then using 
javascript to pass in the `page` via a hidden field in the form.

http://localhost:5000/pagy/docs/how-to/#post-with-page-links

Why are we doing this?

- We need the page number in order to insert it into the search form.
- We are using javascript in order to do this.
- Is it a crazy idea to encode the page number within the link itself as a data-attribute? Then we don't have to bother overriding a controller, we can use javascript to extract the value itself.

```diff
<li class="page-item">
-<a href="/bootstrap_movies/search?page=6&amp;items=2" class="page-link" data-turbo-stream="true">6</a></li>
+<a data-pagy-page-param="6" data-action="pagy#submit" href="/bootstrap_movies/search?page=6&amp;items=2" class="page-link" data-turbo-stream="true" > 6</a></li>
```

Now we can access the page value directly without the need of overriding on the back end. Which is a pain. 

I want this facility because i want to POST and GET to the same URL without touching the back end - but i'm doing this on the pagy-rails. This is not a genuine app. 

EDIT: 

- https://youtu.be/ScxUqW29F7E?t=540 Superails seems to be facing the same problem.

The javascript solution offers **maximum flexibility** without imposing costs. e.g. which comes in handy: for some search results we might want to turbo_stream, and for others we might want to use a turbo_frame, still others we might use a POST request, still others we can use a GET request - without having to change the controller, having the page value directly embedded in a data attribute gives us max flexibility.

```erb
movies/index.html.erb

context-1
<div data-controller="pagy">
  <-- search form -->
<%== pagy_bootstrap_nav(@pagy, link_extra: 'data-turbo-stream="true" data-action: "pagy#submit" ' ) %> (post request) with streams
</div>

context-2
<%== pagy_bootstrap_nav(@pagy) %> (get request) 

context-3
<div data-controller="pagy">
  <-- search form -->
<%== pagy_bootstrap_nav(@pagy, link_extra: 'data-turbo-stream="true" data-action: "pagy#submit" ' ) %> (get request) with streams
</div>

# all to the same or different url
```

@ddnexus  Would be curious to hear your initial ideas here?

**Comment by benkoshy** [comment]
Any thoughts on the above? They're just rough ideas.

**Reply by ddnexus** [reply]
Sorry, I forgot about this.

As a general concept, everything that works better, or even similarly but in another context, could be added to the doc as another way of doing it. (just keep it very very short)

I didn't dig in enough, but it looks like it's a lot more flexible and modern way than the example we  have currently.

**Reply by benkoshy** [reply]
I think it would simplify my code considerably.

I'm yet to think of how to customize the attributes to allow for users to choose who they want to display it: e.g. some might want to different javascript libraries to help them.
